src/Marten/Events/Daemon/HighWater/HighWaterDetector.cs

```diff
@@ -0,0 +1,169 @@
+        private readonly NpgsqlCommand _stateDetection;
+        private readonly NpgsqlParameter _start;
+        private readonly NpgsqlCommand _updateStatus;
+        private readonly NpgsqlParameter _newSeq;
+        private readonly NpgsqlCommand _findSafeSequence;
+        private readonly NpgsqlParameter _safeTimestamp;
+
+        public HighWaterDetector(ITenant tenant, EventGraph graph)
+        {
+            _tenant = tenant;
+
+            _findSafeSequence = new NpgsqlCommand($@"select min(seq_id) from {graph.DatabaseSchemaName}.mt_events where mt_events.timestamp >= :timestamp");
+            _safeTimestamp = _findSafeSequence.AddNamedParameter("timestamp", DateTimeOffset.MinValue);
+
+            _gapDetection = new NpgsqlCommand($@"
+select seq_id
+from   (select
+               seq_id,
+               lead(seq_id)
+               over (order by seq_id) as no
+        from
+               {graph.DatabaseSchemaName}.mt_events where seq_id > :start) ct
+where  no is not null
+  and    no - seq_id > 1
+LIMIT 1;
+select max(seq_id) from {graph.DatabaseSchemaName}.mt_events where seq_id > :start;
+".Trim());
+
+            _start = _gapDetection.AddNamedParameter("start", 0L);
+
+            _stateDetection = new NpgsqlCommand($@"
+select last_value from {graph.DatabaseSchemaName}.mt_events_sequence;
+select last_seq_id, last_updated, transaction_timestamp() as timestamp from {graph.DatabaseSchemaName}.mt_event_progression where name = '{ShardState.HighWaterMark}';
+".Trim());
+
+            _updateStatus =
+                new NpgsqlCommand($"select {graph.DatabaseSchemaName}.mt_mark_event_progression('{ShardState.HighWaterMark}', :seq);");
+            _newSeq = _updateStatus.AddNamedParameter("seq", 0L);
+
+        }
+
+        public async Task<HighWaterStatistics> DetectInSafeZone(DateTimeOffset safeTimestamp, CancellationToken token)
+        {
+            await using var conn = _tenant.OpenConnection();
+
+            var statistics = await loadCurrentStatistics(conn, token);
+
+            _safeTimestamp.Value = safeTimestamp;
+            using (var reader = await conn.ExecuteReaderAsync(_findSafeSequence, token))
+            {
+                if (await reader.ReadAsync(token))
+                {
+                    statistics.SafeStartMark = await reader.GetFieldValueAsync<long>(0, token);
+                }
+            }
+
+            await calculateHighWaterMark(token, statistics, conn);
+
+            return statistics;
+        }
+
+
+        public async Task<HighWaterStatistics> Detect(CancellationToken token)
+        {
+            await using var conn = _tenant.OpenConnection();
+
+            var statistics = await loadCurrentStatistics(conn, token);
+
+
+            await calculateHighWaterMark(token, statistics, conn);
+
+            return statistics;
+        }
+
+        private async Task calculateHighWaterMark(CancellationToken token, HighWaterStatistics statistics,
```

Comment:
Suggestion. I think that it would be good to keep those calculations in the HighWaterStatistics class nad make setters for those properties private. It would be easier to track what's changed where in the future.

Reply:
I completely understand why you're saying that, but I think that's just some extra complexity that would never pay off.